contracts/ProtocolReserve/ProtocolShareReserve.sol

```diff
@@ -92,8 +247,106 @@ contract ProtocolShareReserve is Ownable2StepUpgradeable, ExponentialNoError, Re
      */
     function updateAssetsState(
         address comptroller,
-        address asset
-    ) public override(IProtocolShareReserve, ReserveHelpers) {
-        super.updateAssetsState(comptroller, asset);
+        address asset,
+        IncomeType incomeType
+    ) public override(IProtocolShareReserve) {
+        require(ComptrollerInterface(comptroller).isComptroller(), "ProtocolShareReserve: Comptroller address invalid");
+        require(asset != address(0), "ProtocolShareReserve: Asset address invalid");
+        require(
+            comptroller == corePoolComptroller ||
+                PoolRegistryInterface(poolRegistry).getVTokenForAsset(comptroller, asset) != address(0),
+            "ProtocolShareReserve: The pool doesn't support the asset"
+        );
+
+        Schema schema = Schema.TWO;
+        address vToken = IPrime(prime).vTokenForAsset(asset);
+
+        if (vToken != address(0) && comptroller == corePoolComptroller && incomeType == IncomeType.SPREAD) {
+            schema = Schema.ONE;
+        }
+
+        uint256 currentBalance = IERC20Upgradeable(asset).balanceOf(address(this));
+        uint256 assetReserve = totalAssetReserve[asset];
+
+        if (currentBalance > assetReserve) {
+            uint256 balanceDifference;
+            unchecked {
+                balanceDifference = currentBalance - assetReserve;
+            }
+
+            assetsReserves[comptroller][asset][schema] += balanceDifference;
+            totalAssetReserve[asset] += balanceDifference;
+            emit AssetsReservesUpdated(comptroller, asset, balanceDifference, incomeType, schema);
+        }
+    }
+
+    /**
+     * @dev Fetches the list of prime markets and then accrues interest and
+     * releases the funds to prime for each market
+     */
+    function _accrueAndReleaseFundsToPrime() internal {
+        address[] memory markets = IPrime(prime).allMarkets();
+        for (uint i = 0; i < markets.length; ++i) {
+            address market = markets[i];
+            IPrime(prime).accrueInterest(market);
+            _releaseFund(corePoolComptroller, IVToken(market).underlying(), distributionTargets.length);
+        }
+    }
+
+    /**
+     * @dev Fetches the list of prime markets and then accrues interest
+     * to prime for each market
+     */
+    function _accruePrimeInterest() internal {
+        address[] memory markets = IPrime(prime).allMarkets();
+        for (uint i = 0; i < markets.length; ++i) {
+            address market = markets[i];
+            IPrime(prime).accrueInterest(market);
+        }
+    }
+
+    function _releaseFund(address comptroller, address asset, uint256 totalDistributionTargets) internal {
```

Comment:
I would remove the last parameter. It can be calculated inside the function because it's based on a storage attribute